_posts/2021-10-21-segmentation-torch-android/segmentation-torch-android.Rmd

```diff
@@ -0,0 +1,279 @@
+
+    ...
+    ...
+
+    # Epoch 9/10
+    # Train metrics: Loss: 0.1368                                                           
+    # Valid metrics: Loss: 0.2332
+
+    # Epoch 10/10
+    # Train metrics: Loss: 0.1299                                                           
+    # Valid metrics: Loss: 0.2511
+
+Numbers are just numbers -- how good is the trained model really at segmenting pet images? To find out, we generate segmentation masks for the first eight observations in the validation set, and plot them overlaid on the images. A convenient way to plot an image and superimpose a mask is provided by the `raster` package.
+
+```{r}
+library(raster)
+```
+
+Pixel intensities have to be between zero and one, which is why in the dataset wrapper, we have made it so normalization can be switched off. To plot the actual images, we just instantiate a clone of `valid_ds` that leaves the pixel values unchanged. (The predictions, on the other hand, will still have to be obtained from the original validation set.)
+
+```{r}
+valid_ds_4plot <- pet_dataset(
+  root = dir,
+  split = "valid",
+  size = c(224, 224),
+  normalize = FALSE
+)
+```
+
+Finally, the predictions are generated in a loop, and overlaid over the images one-by-one:
+
+```{r}
+indices <- 1:8
+
+preds <- predict(fitted, dataloader(dataset_subset(valid_ds, indices)))
+
+png("pet_segmentation.png", width = 1200, height = 600, bg = "black")
+
+par(mfcol = c(2, 4), mar = rep(2, 4))
+
+for (i in indices) {
+  
+  mask <- as.array(torch_argmax(preds[i,..], 1)$to(device = "cpu"))
+  mask <- raster::ratify(raster::raster(mask))
+  
+  img <- as.array(valid_ds_4plot[i][[1]]$permute(c(2,3,1)))
+  cond <- img > 0.99999
+  img[cond] <- 0.99999
+  img <- raster::brick(img)
+  
+  # plot image
+  raster::plotRGB(img, scale = 1, asp = 1, margins = TRUE)
+  # overlay mask
+  plot(mask, alpha = 0.4, legend = FALSE, axes = FALSE, add = TRUE)
+  
+}
+```
+
+![Learned segmentation masks, overlaid on images from the validation set.](images/pet_segmentation.png)
+
+Now on to running this model "in the wild" (well, sort of).
+
+## JIT-trace and run on Android
+
+Tracing the trained model will convert it to a form that can be loaded in R-less environments -- for example, from Python, C++, or Java.
+
+<aside>Please see our [introduction to the torch JIT compiler](https://blogs.rstudio.com/ai/posts/2021-08-10-jit-trace-module/).</aside>
+
+We access the `torch` model underlying the fitted `luz` object, and trace it -- where tracing means calling it once, on a sample observation:
+
+```{r}
+m <- fitted$model
+x <- coro::collect(train_dl, 1)
+
+traced <- jit_trace(m, x[[1]]$x)
+```
+
+The traced model could now be saved for use with Python or C++, like so:
+
+```{r}
+traced %>% jit_save("traced_model.pt")
+```
+
+However, since we already know we'd like to deploy it on Android, we directly make use of the specialized function `jit_save_for_mobile()` that, additionally, generates bytecode:
```

Comment:
```suggestion
However, since we already know we'd like to deploy it on Android, we make direct use of the specialized function `jit_save_for_mobile()` that, additionally, generates bytecode:
```

Reply:
yeah, that doesn't really convey what I had in mind ... a good sign that what I originally wrote was a bad choice! Will find something else